packages/@okta/vuepress-site/docs/guides/pwd-optional-widget-sign-in-email/main/java/integrationsteps.md

```diff
@@ -0,0 +1,66 @@
+The following code defines an endpoint that receives the `OTP` and `state` values from the magic link. If `state` is null or `IDX_CLIENT_CONTEXT` is not present in session state, the code assumes the callback has not come from the same browser and prompts the user accordingly.
+
+If successful, the `OTP` and `state` values are passed to the Widget with the rest of its configuration details.
+
+```java
+@GetMapping("/magic-link/callback")
+public ModelAndView handleMagicLinkCallback(
+   HttpServletRequest request,
+   @RequestParam(name = "state") String state,
+   @RequestParam(name = "otp") String otp,
+   HttpSession session) throws MalformedURLException {
+
+   // Check for IDX_CLIENT_CONTEXT in session state and state parameter
+   // If either are null, prompt user to enter code in original browser
+
+   String issuer = oktaOAuth2Properties.getIssuer();
+   String orgUrl = new URL(new URL(issuer), "/").toString();
+
+   ModelAndView mav = new ModelAndView("login");
+   mav.addObject(STATE, state);
+   mav.addObject(OTP, otp);
+   mav.addObject(SCOPES, oktaOAuth2Properties.getScopes());
+   mav.addObject(OKTA_BASE_URL, orgUrl);
+   mav.addObject(OKTA_CLIENT_ID, oktaOAuth2Properties.getClientId());
+   mav.addObject(INTERACTION_HANDLE, idxClientContext.getInteractionHandle());
+   mav.addObject(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+   mav.addObject(CODE_CHALLENGE, idxClientContext.getCodeChallenge());
+   mav.addObject(CODE_CHALLENGE_METHOD, CODE_CHALLENGE_METHOD_VALUE);
+   mav.addObject(REDIRECT_URI,
+      request.getScheme() + "://" + request.getServerName() + ":" +
+      request.getServerPort() + request.getContextPath() + "/authorization-code/callback"
+   );
+   mav.addObject(ISSUER_URI, issuer);
+   session.setAttribute(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+   return mav;
+}
+```
+
+> **Note**: For more information on magic links and OTP, including customizations and complete user journeys, see [Email Magic Links Overview](/docs/guides/email-magic-links-overview/main/).
+
+### 4. Your app handles an authentication success response
+
+After the user successfully verifies their identity, Identity Engine sends an interaction code in a query parameter to `${signInRedirectURI}`. For example, `http://localhost:8080/?interaction_code=2JFmObNY8snovJP6_UK5gI_l7RQ-....`
+
+Create an endpoint that calls your custom authorization server and exchanges the interaction code for access tokens. The user has now signed in.
+
+```java
+@GetMapping("/")
+public String home(
+   @RequestParam(name = "error", required = false) String error,
+   @RequestParam(name = "interaction_code", required = false) String interactionCode,
+   @RequestParam(name = "state", required = false) String state,
+   HttpSession session) {
+
+   if (interactionCode != null && state != null) {
```

Comment:
There is a prompt here but it's not called out in the step's description. What does it do?

Reply:
Not sure what you mean Rob. This is the code that sends the interaction code offf to exchange for an access token.